File: src/clang/magma/compile/ExpandGenerics.c
#include "ExpandGenerics.h"
struct Result_Node_CompileError beforePass(struct State state, struct Node node){if (!node.is()) return new Ok<>(node);

        Node type = node.findNode().orElse(new MapNode());
        if (!type.is()) return new Ok<>(node);

        String value = type.findNodeList()
                .orElse(Lists.empty())
                .get(0)
                .findString()
                .orElse();if (value.equals()) {
            List_<Node> arguments = type.findNodeList().orElseGet(Lists::empty);
            Node param = arguments.get(0);
            Node returns = arguments.get(1);

            return new Ok<>(node.retype()
                    .removeNode()
                    .withNode(, returns)
                    .withNodeList(, Lists.of(param)));
        }if(value.equals()) {
            List_<Node> arguments = type.findNodeList().orElseGet(Lists::empty);
            Node returns = arguments.get(0);

            return new Ok<>(node.retype()
                    .removeNode()
                    .withNode(, returns));
        }

        return new Ok<>(node);
}
struct Result_Node_CompileError afterPass(struct State state, struct Node node){if(node.is()) {
            String value = stringify(node);

            Node symbol = new MapNode().withString(, value);
            Node expansion = new MapNode()
                    .withString(, value)
                    .withNode(, node);

            return new Ok<>(new MapNode()
                    .withNode(, symbol)
                    .withNodeList(, Lists.of(expansion)));
        }

        return new Ok<>(node);
}
struct String stringify(struct Node node){if (node.is()) {
            String caller = node.findNodeList()
                    .orElse(Lists.empty())
                    .stream()
                    .map(element -> element.findString())
                    .flatMap(Streams::fromOption)
                    .collect(new Joiner())
                    .orElse();

            return node.findNodeList().orElse(new JavaList<>())
                    .stream()
                    .map(this::stringify)
                    .collect(new Joiner())
                    .map(arguments -> caller +  + arguments)
                    .orElse(caller);
        }

        return node.findString()
                .or(() -> node.findNode().orElse(new MapNode()).findString())
                .orElse();
}

File: src/java/magma/compile/ExpandGenerics.java
package magma.compile;

import jvm.collect.list.Lists;
import magma.collect.list.List_;
import magma.compile.lang.StringLists;
import magma.compile.transform.State;
import magma.compile.transform.Transformer;
import magma.option.Option;
import magma.option.Some;
import magma.option.Tuple;
import magma.result.Ok;
import magma.result.Result;

public class ExpandGenerics implements Transformer {
    private static Option<List_<String>> qualifyName(State state, List_<String> name) {
        if (name.size() != 1) return new Some<>(name);
        return state.qualifyName(name.findFirst().orElse(""));
    }

    private static List_<String> attachNamespaceToName(State state, List_<String> name) {
        String last = name.findLast().orElse("");
        return state.namespace().add(last);
    }

    @Override
    public Result<Tuple<State, Node>, CompileError> beforePass(State state, Node node) {
        if (node.is("import")) {
            State newState = state.defineImport(node);
            return new Ok<>(new Tuple<>(newState, node));
        }

        if (node.is("qualified")) {
            List_<String> name = StringLists.fromSegments(node);
            List_<String> qualified = qualifyName(state, name)
                    .orElseGet(() -> attachNamespaceToName(state, name));

            Node qualifiedNode = StringLists.toSegments(qualified);
            return new Ok<>(new Tuple<>(state, qualifiedNode));
        }

        if (node.is("definition")) {
            Node type = node.findNode("type").orElse(new MapNode());
            if (type.is("generic")) {
                List_<String> qualifiedName = StringLists.fromSegments(type.findNode("base")
                        .orElse(new MapNode()));

                if (qualifiedName.equalsTo(Lists.of("java", "util", "function", "Function"))) {
                    List_<Node> arguments = type.findNodeList("arguments").orElseGet(Lists::empty);
                    Node param = arguments.get(0);
                    Node returns = arguments.get(1);

                    Node definition = node.retype("functional-definition")
                            .removeNode("type")
                            .withNode("return", returns)
                            .withNodeList("params", Lists.of(param));

                    return new Ok<>(new Tuple<>(state, definition));
                }

                if (qualifiedName.equalsTo(Lists.of("java", "util", "function", "Supplier"))) {
                    List_<Node> arguments = type.findNodeList("arguments").orElseGet(Lists::empty);
                    Node returns = arguments.get(0);

                    Node definition = node.retype("functional-definition")
                            .removeNode("type")
                            .withNode("return", returns);

                    return new Ok<>(new Tuple<>(state, definition));
                }

                return new Ok<>(new Tuple<>(state, node));
            }
        }

        return new Ok<>(new Tuple<>(state, node));
    }

    @Override
    public Result<Tuple<State, Node>, CompileError> afterPass(State state, Node node) {
        if (node.is("root")) {
            return new Ok<>(new Tuple<>(state.clearImports(), node));
        }

        if (node.is("generic")) {
            Node expansionGroup = new MapNode("group")
                    .withNode("child", node)
                    .withNodeList("expansions", Lists.of(node.retype("expansion")));

            return new Ok<>(new Tuple<>(state, expansionGroup));
        }

        return new Ok<>(new Tuple<>(state, node));
    }
}
